doc: Netboot 5340 doc update (#4382)
nrfconnect/sdk-nrf: samples/nrf5340/netboot/README.rst

.. _nc_bootloader:

> nRF5340: Network Core Bootloader

Author: @ru-fu I know this, but the issue here is that B0 does have a proper name (Nordic Secure Immutable Bootloader) while B0n, mentioned here, does not, but think it should. I wanted just to see how it looked like. I can eventually revert it back till an official name is decided by PMT.

> nRF5340: Network Core Bootloader

Author: Reverted for now.

################################

.. contents::
   :local:
   :depth: 2

> The Network Core Bootloader sample, also called B0n, is an immutable first-stage bootloader that can update the application firmware on the network core of the nRF5340 System on Chip (SoC).

Author: @ru-fu Please read my previous comment.

> The Network Core Bootloader sample, also called B0n, is an immutable first-stage bootloader that can update the application firmware on the network core of the nRF5340 System on Chip (SoC).

Author: as mentioned in the previous comment, reverted for now.

Overview
********

> The Network Core Bootloader sample protects the flash memory areas allocated to both itself and the application running on the network core.

Author: @ru-fu there are 2 main purposes, to my understanding: protecting the flash and allowing for DFU of the netcore app. If I understood correctly, B0n is not required to just start an image on the netcore. @hakonfam Could you please clarify this?

> The Network Core Bootloader sample protects the flash memory areas allocated to both itself and the application running on the network core.

Reviewer: You don't need a bootloader to start the application. If you remove the bootloader the application would be placed at the beginning of memory, and would be executed directly when the network core is powered on. It does not make sens IMO to say that the main purpose of the bootloader is to start the application.

You must use this sample as a child image of a :ref:`multi-image <ug_multi_image>` build, where MCUboot is enabled and there is a network core application.
MCUboot verifies and shares with the network core bootloader any new network core application image received through a device firmware update (DFU) transport layer, like a serial or a Bluetooth LE connection.
For this reason, without MCUboot, this sample does nothing other than launching directly the application.

During the boot process, the Network Core Bootloader sample and MCUboot interact as follows:

> 1. MCUboot writes the Network Core Bootloader sample configuration to the shared SRAM.

Reviewer: MCUboot only writes a _command_ to the Network Core Bootloader if it identifies a network core application image. MCUboot does not write anything to shared SRAM if no network core update is found.

Author: fixed

     In case it received a new network core application image, the configuration includes an update instruction and both the SHA and the data of the image.
#. The bootloader sample locks the flash memory areas containing itself and its configuration.
     Locking is done using the ACL peripheral.
     For details on locking, see the :ref:`fprotect_readme` driver.

> #. The sample performs any pending network core firmware update.

Reviewer: This image is now called "Network Core Bootloader" "The bootloader sample" and "The sample" - would it be better if we were consistent at least in this list?

Author: As mentioned in the other comment, fixed

     The sample calls the :ref:`subsys_pcd` library to inspect the SRAM region shared with the application core:

     a. If the application core has written an update instruction, the bootloader copies the specified data range to the application partition on the network core.

>      #. Once the copy is done, the sample compares the SHA of the data in the application partition against the SHA specified in the shared SRAM.

Reviewer: Again, "the sample" here, but "the bootloader" above.

Author: Well, I didn't want to repeat the bootloader all the time. Anyway, I changed these mentions in the entire section. Fixed.

     #. It then communicates the result of the comparison to MCUboot using the shared SRAM.
#. It then locks the flash memory areas containing the network core application.
     Locking is done using the ACL peripheral.
     For details on locking, see the :ref:`fprotect_readme` driver.
#. It boots the application on the network core.
     After possibly performing a firmware update, and enabling flash memory protection, the network core bootloader uninitializes all peripherals that it used and boots the application.

Requirements
************

The sample supports the following development kit:

.. table-from-rows:: /includes/sample_board_rows.txt
   :header: heading
   :rows: nrf5340dk_nrf5340_cpunet

.. _net_bootloader_build_and_run:

Building and running
********************

Follow the steps below to include the sample as a child image in a :ref:`multi-image <ug_multi_image>` build that contains a network core application:

#. Add MCUboot to the build by enabling the :option:`CONFIG_BOOTLOADER_MCUBOOT` option in the application that runs on the application core.

>    When doing so, the build system includes the sample in the build by enabling automatically the :option:`CONFIG_SECURE_BOOT` option for the application that runs on the network core.

Author: fixed

#. Enable the :ref:`subsys_pcd` library for MCUboot by setting the :option:`CONFIG_PCD` option when building its image.

The build system generates a new set of firmware update files.
These files match the ones described in :ref:`mcuboot:mcuboot_ncs`, except that they contain the network core application firmware and are prefixed with ``net_core_``.

> This sample can be found under :file:`samples/nrf5340/netboot/` in the |NCS| folder structure.

Author: fixed

     Locking is done using the ACL peripheral.
     For details on locking, see the :ref:`fprotect_readme` driver.
#. It performs any pending network core firmware update.
     The sample calls the :ref:`subsys_pcd` library to inspect the SRAM region shared with the application core:

     a. If the application core has written an update instruction, the bootloader copies the specified data range to the application partition on the network core.
     #. Once the copy is done, the sample compares the SHA of the data in the application partition against the SHA specified in the shared SRAM.
     #. It then communicates the result of the comparison to the application core through the shared SRAM.
#. It then locks the flash memory areas containing the application.
     Locking is done using the ACL peripheral.
     For details on locking, see the :ref:`fprotect_readme` driver.
#. It boots the application on the network core.
     After possibly performing a firmware update, and enabling flash memory protection, the network core bootloader uninitializes all peripherals that it used and boots the application.

Requirements
************

> The sample supports the following development kits:

Reviewer: It only supports the `nrf5340dk_nrf5340_cpunet` board, the table below correctly lists that build target as the only supported, but maybe this sentence should reflect that as well?

Author: good point, fixed

.. table-from-rows:: /includes/sample_board_rows.txt
   :header: heading
   :rows: nrf5340dk_nrf5340_cpunet

.. _net_bootloader_build_and_run:

Building and running
********************

The recommended way of building the sample is to include it as a child image in a :ref:`multi-image <ug_multi_image>` build where there is an application that runs on the network core, rather than building it stand-alone.
There are two ways to include it in a multi-image build:

> * If you don't need the capability to update the sample, enable the :option:`CONFIG_SECURE_BOOT` option in the application that runs on the network core.

Reviewer: This point can be removed IMO.

This sample gives no value if MCUboot is not enabled on the app core. This because this sample is solely driven from the configuration written to shared SRAM by MCUboot. If MCUboot is not in place, this sample does nothing other than jumping to the application.

Author: fixed

* If you need the capability to update the sample, enable the :option:`CONFIG_BOOTLOADER_MCUBOOT` option in the application that runs on the application core to add MCUboot to the build.
  When doing so, the build system includes the sample in the build by enabling automatically the :option:`CONFIG_SECURE_BOOT` option for the application that runs on the network core.
  In addition, you must also enable the :ref:`subsys_pcd` library for MCUboot by setting the :option:`CONFIG_PCD` option when building its image.
  See :ref:`ug_multi_image` for more information about setting configurations to a specific image.

  .. note::
     In this setup, MCUboot verifies any new image received through a device firmware update (DFU) transport layer, like a serial or a Bluetooth LE connection, for both the application and the network core.

>      This includes the network core bootloader image.

Reviewer: No, the network core bootloader image is never verified. It is immutable and does not need to be verified.

Author: fixed

     See :ref:`subsys_pcd` for more details.

When you include this sample in a multi-image build, the build system generates a new set of firmware update files.
These files match the ones described in :ref:`mcuboot:mcuboot_ncs`, except that they contain the network core application firmware and are prefixed with ``net_core_``.

This sample can be found under :file:`samples/nrf5340/netboot/` in the |NCS| folder structure.

See :ref:`configure_application` for information on how to enable the required configuration options.
Then follow the instructions in :ref:`ug_nrf5340_building` to build and program the images for the network and application core.

.. note::
   To try out the network core bootloader sample, you can use the :ref:`peripheral_uart` sample as the basis for the multi-image build.
   This sample automatically includes the network core sample :ref:`bluetooth-hci-rpmsg-sample` when built for the nRF5340 DK.

>    You can then apply the options mentioned above to include the network core sample with or without MCUboot.

Reviewer: The sample is only included if MCUboot is enabled for the `peripheral_uart` sample.

Reviewer 2: The `peripheral_uart` is the only sample that the netboot sample can be applied with?

Reviewer: No, my point was that for any app core sample you have to enable MCUboot for the netboot bootloader to be included.
In addition to this any app core sample must also include a net core child image (which `peripheral_uart` does)

> Testing

Reviewer: We should have the build instructions in place both for command line and SES

Author: @hakonfam the whole scope of the JIRA that made this PR necessary was to actually make the sample docs mention the official way we build samples in NCS, through SES. The instructions are linking to the pages that address this. In a previous version of this doc I realized it would be just a copypaste of these, so to remove any duplication I linked them there like it is done in most of the other NCS samples.

   :local:
   :depth: 2

This bootloader sample implements an immutable first-stage bootloader that can update the application firmware on the network core of the nRF5340 System on Chip (SoC).

> In addition to this, the network core bootloader sample performs flash memory protection of both itself and the application.

Author: Rephrased, it should be slightly better now.

Overview
********

> The network core bootloader sample supports transporting an already verified and authenticated firmware upgrade from the application core flash memory to the network core flash memory, as well as providing flash memory protection.

Author: ok, let's stick to how it is usually used in other ncs docs. Fixed.

This is accomplished by the following steps:

1. Locking the flash memory areas containing the bootloader.
     The bootloader sample locks the flash memory areas that contain the sample bootloader and its configuration.
     Locking is done using the ACL peripheral.
     For details on locking, see the :ref:`fprotect_readme` driver.

#. Performing any pending network core firmware upgrade.
     The sample invokes the :ref:`subsys_pcd` library to inspect an SRAM region shared with the application core.
     If the application core has written an update instruction, the bootloader copies the specified data range to the application partition on the network core.

>      Once the copy is done, it compares the SHA of the data in the application partition against the SHA specified in the shared SRAM.

Author: fixed

     It then communicates the result of the comparison to the application core through the shared SRAM.

#. Locking the flash memory areas containing the application.
     The sample locks the flash memory areas that contain the application.
     Locking is done using the ACL peripheral.
     For details on locking, see the :ref:`fprotect_readme` driver.

#. Booting the application on the network core.
     After possibly performing a firmware update, and enabling flash memory protection, the network core bootloader uninitializes all peripherals that it used and boots the application.

The most common use case for the network core bootloader sample is to be included as a child image in a multi-image build, rather than being built stand-alone.

Requirements
************

The sample supports the following development kits:

.. table-from-rows:: /includes/sample_board_rows.txt
   :header: heading
   :rows: nrf5340dk_nrf5340_cpunet

.. _net_bootloader_build_and_run:

Building and running
********************

The recommended way of building the sample is to use the multi-image feature of the build system.

> You can include the sample in your multi-image build as follows:

Author: You can choose either of the 2 options. I rephrased the introductory sentence.

> * Enabling the :option:`CONFIG_SECURE_BOOT` option in the application that has to run on the nRF5340 network core.

Author: fixed

>   The build system will include this sample automatically.

Author: fixed

* Enabling the :option:`CONFIG_BOOTLOADER_MCUBOOT` and :option:`CONFIG_BT_RPMSG_NRF53` options in the application that has to run on the application core.

>   The build system will enable the :option:`CONFIG_SECURE_BOOT` option automatically for the network core application.

Author: With these 2 options set, fixed

When you include this image in a multi-image build, the build system generates a new set of firmware update files
These files match the ones described in :ref:`mcuboot:mcuboot_ncs`, except that they contain the network core application firmware and are prefixed with ``net_core_``.

.. |sample path| replace:: :file:`samples/nrf5340/netboot/`

.. include:: /includes/build_and_run.txt
See also the following pages:

* :ref:`ug_nrf5340_building` for additional details on building samples on the nRF5340.
* :ref:`configure_application` for additional details on configuring your application.

See below for an example of a multi-image build that includes :ref:`peripheral_uart`, using either SEGGER Embedded Studio or the command line:

.. tabs::

   .. group-tab:: SEGGER Embedded Studio

>       Follow the instructions linked above, enabling the following options:

Author: As I changed the way the doc is structured, this part is now missing.

      * :option:`CONFIG_BOOTLOADER_MCUBOOT`:

        #. Select :guilabel:`Project` -> :guilabel:`Configure nRF Connect SDK project`.

>         #. Go to :guilabel:`Modules` -> :guilabel:`Nordic nRF Connect` -> :guilabel:`Bootloader` and set :guilabel:`Use MCUboot` to enable :option:`CONFIG_BOOTLOADER_MCUBOOT`.

Reviewer 3: This isn't where this option is located when I test it. Please check.

Author: fixed

        #. Click :guilabel:`Configure`.

>       * :option:`mcuboot_CONFIG_PCD`:

Reviewer 3: This is not an option. You want to configure option CONFIG_PCD for the mcuboot image.

Author: fixed

        #. Select :guilabel:`Project` -> :guilabel:`Configure nRF Connect SDK project`.

>         #. Go to :guilabel:`Modules` -> :guilabel:`Nordic nRF Connect` -> :guilabel:`Peripheral CPU DFU (PCD)` and set :guilabel:`Use Peripheral CPU DFU (PCD)` to enable :option:`mcuboot_CONFIG_PCD`.

Reviewer 3: ```suggestion
        #. Go to :guilabel:`Modules` -> :guilabel:`Nordic nRF Connect` -> :guilabel:`Peripheral CPU DFU (PCD)` and select :guilabel:`Peripheral CPU DFU Support (PCD)` to enable :option:`CONFIG_PCD`.
```

Reviewer 3: Not sure why this shows up different for me. Please test.

Reviewer 3: This option is already enabled for me. So maybe better to say that users should make sure it is enabled?

Author: fixed

        #. Click :guilabel:`Configure`.

      The :ref:`peripheral_uart` sample will automatically include the network core sample :ref:`bluetooth-hci-rpmsg-sample` when built for the nRF5340 DK.
      Enabling MCUboot in the application core automatically includes the network core bootloader.

>       In addition to this, you must also enable the :ref:`subsys_pcd` library for the MCUboot image.

Reviewer 3: Maybe put this at the top so you know what you're doing before you start?

> 

Reviewer 3: Now you have configured the sample, but you haven't built or programmed it.

Author: As I changed the way the doc is structured, this part is now missing.

   .. group-tab:: Command Line

      #. |connect_terminal|

>          Note that the nRF5340 DK has multiple UART instances, so the correct port must be identified.

Reviewer 2: This would be an ideal place to link the [doc that specifices these COM ports](https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/ug_nrf5340.html#getting-logging-output), as this statement is not very helpful without this info.

Reviewer 3: Actually, this step can be removed here, since you have it in the testing section below. So you should move the additional information to the testing section.

Author: I added the mention to the COM ports, as this step is actually needed for both for programming the sample to the SoC and, later, for testing.

> 

Reviewer 2: I recommend refactoring this step-by-step procedure by having a brief description as the bullet, followed by a code block of the command, then followed by its additional information.

* Steps are then succinctly described rather than described by dissecting the command.
* Aligns with most other bullet/number lists across our documentation.
* Allows these commands to make use of restructuredText's `code-block` or `parsed-literal` directives.

Author: fixed

>       #. ``west build -b nrf5340dk_nrf5340_cpuapp -d build_netboot samples/bluetooth/peripheral_uart -f -- -DCONFIG_BOOTLOADER_MCUBOOT=y -Dmcuboot_CONFIG_PCD=y``

Reviewer 2: I suggest removing `-f` from this command. It is not necessary to compile the sample, and it is probably best to avoid encouraging force options unless they are necessary for a specific circumstance.

Author: fixed

Author: fixed

         The :ref:`peripheral_uart` sample will automatically include the network core sample :ref:`bluetooth-hci-rpmsg-sample` when built for the nRF5340 DK.

>          Enabling MCUboot (``-DCONFIG_BOOTLOADER_MCUBOOT=y``) in the application core automatically includes the network core bootloader.

Reviewer 2: ```suggestion
         Enabling MCUboot  in the application core by compiling with the :option:`CONFIG_BOOTLOADER_MCUBOOT` option automatically includes the network core bootloader.
```

Author: fixed

Author: fixed

>          In addition to this, you must also enable the :ref:`subsys_pcd` library for the MCUboot image (``-Dmcuboot_CONFIG_PCD=y``).

Reviewer 2: ```suggestion
         In addition to this, you must also enable the :ref:`subsys_pcd` library for the MCUboot image by compiling it with the :option:`CONFIG_PCD` option.
```

Author: fixed

Author: fixed

      #. ``west flash -d build_netboot``

>          Program the merged hex files for both domains (application core and network core).

Reviewer 3: HEX files or hexadecimal files

   :local:
   :depth: 2

The network core bootloader sample implements an immutable first-stage bootloader that can update the application firmware on the network core of the nRF5340 System on Chip (SoC).

Overview
********

This sample supports transporting an already verified and authenticated firmware update from the application core flash memory to the network core flash memory.
Also, this sample protects the flash memory areas allocated to both itself and the application running on the network core.

> The bootloader accomplishes these objectives as follows:

Reviewer 3: ```suggestion
The bootloader accomplishes the following objectives:
```

Author: What I meant here is that the 2 previously listed objectives are accomplished by the following sequence of actions. I'll rephrase it.

Author: I rephrased the entire section a bit, the reason for line 18 should now be more clear.

1. Locking the flash memory areas containing the bootloader.
     The bootloader sample locks the flash memory areas that contain the sample bootloader and its configuration.
     Locking is done using the ACL peripheral.
     For details on locking, see the :ref:`fprotect_readme` driver.

#. Performing any pending network core firmware update.
     The sample calls the :ref:`subsys_pcd` library to inspect the SRAM region shared with the application core:

     a. If the application core has written an update instruction, the bootloader copies the specified data range to the application partition on the network core.
     #. Once the copy is done, the sample compares the SHA of the data in the application partition against the SHA specified in the shared SRAM.
     #. It then communicates the result of the comparison to the application core through the shared SRAM.

#. Locking the flash memory areas containing the application.

>      The sample locks the flash memory areas that contain the application.

Reviewer 3: This sentence is exactly the same as the heading - no need for duplication. If you have more information to add, you should extend the second sentence; otherwise just get rid of it.
If you want to be parallel to the first step, you could extend the heading for the first step ("... containing the bootloader and its configuration") and get rid of the duplication there as well.

Author: fair enought, fixed

     Locking is done using the ACL peripheral.
     For details on locking, see the :ref:`fprotect_readme` driver.

#. Booting the application on the network core.
     After possibly performing a firmware update, and enabling flash memory protection, the network core bootloader uninitializes all peripherals that it used and boots the application.

The most common use case for the network core bootloader sample is to be included as a child image in a multi-image build, rather than being built stand-alone.

Requirements
************

The sample supports the following development kits:

.. table-from-rows:: /includes/sample_board_rows.txt
   :header: heading
   :rows: nrf5340dk_nrf5340_cpunet

.. _net_bootloader_build_and_run:

Building and running
********************

The recommended way of building the sample is to use the :ref:`multi-image feature <ug_multi_image>` of the build system.

> You can include the sample in your multi-image build in two different ways:

Reviewer 3: How do you choose which option to use?
It seems to depend on if you have an application on the application core, or if you want to include the bootloader in the application from the network core or the application core ...?

Once clarified, it might be clearer to use a definition list than a bullet list, so something like:

```
To include the bootloader from the network core application
   Enable ...

To include the bootloader from the application core application
   Enable ...
```

Author: @ru-fu I rephrased the section, please check if you still think that a definition list might be better than a bullet list.

Author: outdated

> * Enabling the :option:`CONFIG_SECURE_BOOT` option in the application that runs on the nRF5340 network core.

Author: Fixed

  With this option enabled, the build system includes the sample automatically.

> * Enabling the :option:`CONFIG_BOOTLOADER_MCUBOOT` and :option:`CONFIG_BT_RPMSG_NRF53` options in the application that runs on the application core.

Reviewer 2: Users don't need to give `CONFIG_BT_RPMSG_NRF53` themselves - this is included automatically when compiling for netboot.

The only options required for the user to give in order to build this sample are `mcuboot_CONFIG_PCD=y` and `CONFIG_BOOTLOOADER_MCUBOOT=y`

Author: fixed

  In a multi-image build where there is also an application running on the network core, enabling MCUboot in the application core automatically includes the network core bootloader.
  As such, with these options enabled, the build system automatically enables the :option:`CONFIG_SECURE_BOOT` option for the network core application.

>   In addition, you must also enable the :ref:`subsys_pcd` library for the MCUboot image, enabling the option :option:`CONFIG_PCD`.

Reviewer 2: ```suggestion
  In addition, you must also enable the :ref:`subsys_pcd` library for MCUboot by setting the :option:`CONFIG_PCD` option when building its image.
  See :ref:`ug_multi_image` for more information about setting configurations to a specific image.
```

Author: fixed

  .. note::

     The firmware is signed exactly like the application firmware is signed.

>      As such, the firmware is also verified by MCUboot on the application core.

Reviewer 3: What does "as such" mean?

Author: rephrased

     See :ref:`subsys_pcd` for more details.

> When you include this image in a multi-image build, the build system generates a new set of firmware update files

Reviewer 3: ```suggestion
When you include this image in a multi-image build, the build system generates a new set of firmware update files.
```

Author: fixed

  In a multi-image build where there is also an application running on the network core, enabling MCUboot in the application core automatically includes the network core bootloader.
  As such, with these options enabled, the build system automatically enables the :option:`CONFIG_SECURE_BOOT` option for the network core application.

Testing
=======
  In addition, you must also enable the :ref:`subsys_pcd` library for the MCUboot image, enabling the option :option:`CONFIG_PCD`.

To test the network core bootloader sample run the following commands:
  .. note::

#. |connect_terminal|
     The firmware is signed exactly like the application firmware is signed.
     As such, the firmware is also verified by MCUboot on the application core.
     See :ref:`subsys_pcd` for more details.

   Note that on the nRF5340 DK has multiple UART instances, so the correct port must be identified.
When you include this image in a multi-image build, the build system generates a new set of firmware update files
These files match the ones described in :ref:`mcuboot:mcuboot_ncs`, except that they contain the network core application firmware and are prefixed with ``net_core_``.

> #. ``west build -b nrf5340dk_nrf5340_cpuapp -d build_netboot samples/bluetooth/peripheral_uart -f -- -DCONFIG_BOOTLOADER_MCUBOOT=y -Dmcuboot_CONFIG_PCD=y``

Reviewer 2: I like how concise these changes are, but I would suggest mentioning the `samples/bluetooth/peripheral_uart` sample as the basis for the multi-image build for testing this. It's a great sample to use for netboot, and useful to know about for users that are just starting to get familiar with netboot on the 5340.

Author: @kyberdin added a mention in a note

> * :ref:`configure_application` for additional details on configuring your application.

Reviewer 3: This feels a bit like throwing random information at the user and telling them to figure out how to do it. ;)
You don't need to include the general instructions for building and running if we have specific ones for nRF5340. And we haven't told users anything about configuring yet, so these are not "additional details".

So instead something like:
"See :ref:`configure_application` for information on how to enable the required configuration option.
Then follow the instructions in :ref:`ug_nrf5340_building` to build and program the images for the network and application core."

Author: Right, I was unsure if I could get rid of the include. Fixed

>       #. Observe that the output includes the following:

Reviewer 3: ```suggestion
      #. Observe that the output includes the following line:
```

Author: fixed
